packages/tsutils/src/typeGuards.ts

```diff
@@ -8,5 +8,13 @@ export const isDefined = <T>(value: T): value is Exclude<T, undefined> => value
 export const isNotNullish = <T>(val: T): val is Exclude<T, undefined | null> =>
   val !== undefined && val !== null;
 
+export const ensure = <T>(val: T) => {
+  if (isNotNullish(val)) {
+    return val;
+  }
+
+  throw new Error(`Unexpected null or undefined value`);
+};
```

Comment:
This crops up pretty often that we want a typeguard that ensures that a value is not null or undefined. Took the name from this from a similar thread on Stack Overflow.

Reply:
Hey Rohan, if you don’t mind a bit of an intern question, did you consider an assertion here, and if so, why `const bar = ensure(foo);` over a usage more like this?

```ts
assertIsDefined(foo);
const bar = foo;
```

where:

```ts
function assertIsDefined<T>(val: T): asserts val is NonNullable<T> {
  if (val === undefined || val === null) {
    throw new AssertionError('Unexpected null or undefined value');
  }
}
```

I’m obviously less familiar with how these values are used in our code, but from what I understand, the typing system will know that the value is non-nullish after the `assertIsDefined` call

Don’t worry if you’ve got too much on your plate to answer! 🙂

Reply:
Great question! I hadn't considered an assertion, so thanks for mentioning it. I think both approaches have their place.

I like assertions when we're working with variables or arguments. eg.
```
const lowerCase = (val: string | null) => {
  assertIsDefined(val); // Clean way to validate args at the start of a function
  return val.toLowerCase();
};
```
whereas an `ensure` function like this is quite useful when we're doing inline operations or chaining, eg.
```
const commonWord = (arr1: string[], arr2: string[]) => {
  return ensure(arr1.find(val => arr2.includes(val)));
};

const lowerCaseAll = (vals: (string | null)[]) => {
  return vals.map(ensure).map(val => val.toLowerCase());
};
```

I do like the assertion approach often though as it feels a bit more clear what's going on (though possibly that's also due to `ensure` being a bit vaguely named by me...).

So it's all just code style really, did you have any thoughts?

Reply:
Oh, yep, that makes a lot of sense; and having had a squiz at the usages of `ensure` in this PR, I do feel that using an assertion instead might be slightly clunkier in this case.

I agree that `ensure` isn’t the most transparent name, but (1) I don’t have a better alternative; and (2) it’s pretty simple to grasp.

At the risk of getting needlessly in the weeds for a non-issue, I suppose my instinct then would be:

```ts
const ensure(val) => {
  assertIsDefined(val);
  return val;
}
```

Though, I missed earlier that you’re using `Exclude<>`, so these two approaches are well and truly equivalent 😅

Thank you for the detailed response—particularly the concrete examples! 🙏

Reply:
There's always a bit of fun in getting into the weeds :relaxed: thanks for raising some interesting questions. Yep, the `Exclude` achieves the same effect as the `NonNullable` assertion, though I suppose we should use the built in type. So I might change that.

And your instincts on how `ensure` should read feel a little nicer too (assert, if assertion passes then return) so I'll change that as well :+1: